### PR TITLE
Add batched exports to the Prometheus Remote Write Exporter 

### PR DESCRIPTION
**Pull request by JasonXZLiu**
**Description:** 

Currently, there is no limit on the size of the remote_write export request to a Cortex instance. This could create problems if one Prometheus scrape produces an extreme amount of timeseries metrics. 

This PR creates a batch of concurrent requests where the max size of a request is 3MB and the max number of timeseries per request is 40,000.

A future PR will tackle making the max size of each request configurable.

**Testing:** 

* Unit tests were fixed
* Performance tests were done with 15000 Prometheus metrics (around 100k timeseries) per scrape. The first and last metric of each Prometheus metric type were all found.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@huyan0 will you be able to review this as the original author of the exporter?

**Comment by JasonXZLiu** [comment]
@amanbrar1999 is actually the current maintainer from AWS for the Prometheus Remote Write Exporter. Would also be great if we could get a review from @jmacd. Thank you!

cc - @alolita

**Review by amanbrar1999 — APPROVED** [review]
LGTM

**Review by huyan0 — APPROVED** [review]

**Review by rakyll — APPROVED** [review]

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/exporter.go:278]: Nit: if len(requests)<maxConcurrentRequests then spawning maxConcurrentRequests goroutines is excessive.

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/exporter.go:283]: Nit: alternatively, these goroutines could be started once when the is Started instead of doing it every time export is called, which may add a bit of overhead if the export requests are tiny.

File: exporter/prometheusremotewriteexporter/helper.go
Comment (tigrannajaryan) [inline exporter/prometheusremotewriteexporter/helper.go:244]: This may produce an empty request if len(tsArray)==0. Is that OK in `execute`?

File: exporter/prometheusremotewriteexporter/exporter.go
Comment (JasonXZLiu) [reply exporter/prometheusremotewriteexporter/exporter.go:283]: Hmm, I added the limit you mentioned before (about spawning the minimum number of goroutines necessary), so I don't think we would have as much overhead as before. You're right that it would be better if we started these goroutines when the exporter itself is started. We'll create an issue for this and address it in a separate PR.

Comment (rakyll) [reply exporter/prometheusremotewriteexporter/exporter.go:285]: @tigrannajaryan I wonder if it's your experience that they are adding more overhead. It's so cheap to start new goroutines (even though not cost-free) that I've seen it relatively rare in Go to reuse goroutines. It's one of the goals of the goroutines to be cheap so the devs don't care about goroutine pools, etc. We might add this as a TODO to take a look at but given how net/http spawns gorotuines each time, I don't think this is an improvement worth investigating right now.

Comment (tigrannajaryan) [reply exporter/prometheusremotewriteexporter/exporter.go:285]: It is likely not going to be a measurable difference for typical batch size. Was more of a though exercise, not an ask to change anything. :-)
What you have is fine.

**Review by tigrannajaryan — APPROVED** [review]
LGTM

**Pull request merged**
